File: haskell/src/Messages.hs
{-# LANGUAGE DeriveGeneric,
             DuplicateRecordFields #-}

module Messages (
  Direction (
    Up,
    Down
  ), 
  Event (
    Initialize,
    ElevatorBroken,
    ElevatorDoorBroken,
    FloorDoorBroken,
    FloorPassed,
    ElevatorStopped,
    ElevatorDoorOpened,
    ElevatorDoorClosed,
    FloorDoorOpened,
    FloorDoorClosed,
    ElevatorButtonPressed,
    FloorButtonPressed
  ), 
  Command (
    MoveElevator,
    StopElevator,
    OpenElevatorDoor,
    CloseElevatorDoor,
    OpenFloorDoor,
    CloseFloorDoor,
    SetElevatorButtonLight,
    SetFloorButtonLight
  )
) where

import GHC.Generics

data Direction = Up | Down deriving (Generic, Show)

data Event = Initialize { floors :: Int, shafts :: Int, capacity :: Int }
           | ElevatorBroken { shaft :: Int, reason :: String }
           | ElevatorDoorBroken { shaft :: Int, reason :: String }
           | FloorDoorBroken { floor :: Int, shaft :: Int, reason :: String }
           | FloorPassed { shaft :: Int }
           | ElevatorStopped { shaft :: Int }
           | ElevatorDoorOpened { shaft :: Int }
           | ElevatorDoorClosed { shaft :: Int }
           | FloorDoorOpened { floor :: Int, shaft :: Int }
           | FloorDoorClosed { floor :: Int, shaft :: Int }
           | ElevatorButtonPressed { shaft :: Int, floor :: Int }                   
           | FloorButtonPressed { floor :: Int, direction :: Direction } deriving (Show, Generic)

data Command = MoveElevator { shaft :: Int, direction :: Direction }
             | StopElevator { shaft :: Int }
             | OpenElevatorDoor { shaft :: Int }
             | CloseElevatorDoor { shaft :: Int }
             | OpenFloorDoor { floor :: Int, shaft :: Int }
             | CloseFloorDoor { floor :: Int, shaft :: Int }
             | SetElevatorButtonLight { shaft :: Int, floor :: Int, on :: Bool }
             | SetFloorButtonLight { floor :: Int, direction :: Direction, on :: Bool } deriving (Show, Generic)

File: scala/src/main/scala/Messages.scala


sealed trait Direction
object Direction {
  case object Up extends Direction
  case object Down extends Direction
}

sealed trait Event
object Event {
  case class Initialize(floors: Int, shafts: Int, capacity: Int) extends Event
  case class ElevatorBroken(shaft: Int, reason: String) extends Event
  case class ElevatorDoorBroken(shaft: Int, reason: String) extends Event
  case class FloorDoorBroken(floor: Int, shaft: Int, reason: String) extends Event
  case class FloorPassed(shaft: Int) extends Event
  case class ElevatorStopped(shaft: Int) extends Event
  case class ElevatorDoorOpened(shaft: Int) extends Event
  case class ElevatorDoorClosed(shaft: Int) extends Event
  case class FloorDoorOpened(floor: Int, shaft: Int) extends Event
  case class FloorDoorClosed(floor: Int, shaft: Int) extends Event
  case class ElevatorButtonPressed(shaft: Int, floor: Int) extends Event
  case class FloorButtonPressed(floor: Int, direction: Direction) extends Event
}

sealed trait Command
object Command {
  case class MoveElevator(shaft: Int, direction: Direction) extends Command
  case class StopElevator(shaft: Int) extends Command
  case class OpenElevatorDoor(shaft: Int) extends Command
  case class CloseElevatorDoor(shaft: Int) extends Command
  case class OpenFloorDoor(floor: Int, shaft: Int) extends Command
  case class CloseFloorDoor(floor: Int, shaft: Int) extends Command
  case class SetElevatorButtonLight(shaft: Int, floor: Int, on: Boolean) extends Command
  case class SetFloorButtonLight(floor: Int, direction: Direction, on: Boolean) extends Command
}
